lib/net/ssh/transport/session.rb

```diff
@@ -64,7 +64,13 @@ def initialize(host, options={})
 
       debug { "establishing connection to #{@host}:#{@port}" }
       factory = options[:proxy] || TCPSocket
-      @socket = timeout(options[:timeout] || 0) { @bind_address.nil? || options[:proxy] ? factory.open(@host, @port) : factory.open(@host,@port,@bind_address) }
+      @socket = timeout(options[:timeout] || 0) {
+        case
+        when options[:proxy] then factory.open(@host, @port, options)
```

Comment:
as noted by https://github.com/net-ssh/net-ssh/issues/220 ```options[:proxy]``` can't be both a factory class and dictionary. 
```factory = options[:proxy] || TCPSocket```
```factory.open(@host, @port, options)```